### ISSUE TITLE
How to generate initial function embedding json

### ISSUE THREAD

**Issue by TakahiroHaruyama** [wangyongpang/BinEnhance#2]
Hi, thanks for sharing your research implementation!

I'm trying to reproduce the training and inference part. First I need to run whitening_transformation.py to reduce the dimension. The code takes input-dir as an argument and it's actually a json file.
On the other hand, HermesSim (one of the baselines) generates function embeddings in PyTorch tensor format (.pkl).

```
In [4]: pkl_data = pickle.load(open('dataset/Dataset-1/testing_Dataset-1.pkl', 'rb'))

In [5]: type(pkl_data)
Out[5]: torch.Tensor

In [6]: pkl_data.shape
Out[6]: torch.Size([522003, 384])
```

So how can we translate the .pkl to .json? The whitening_transformation.py likely expects a dictionary with a function id as the key and embedding data as the value. Is the key string format the same as test_function_embeddings.json in Eval data (binary_file|||function_name)? 
And if you wrote a script for the conversion, could you share it?

**Comment by TakahiroHaruyama** [comment]
I wrote the conversion script. Please let me know if the key format is not correct.

```
#!/usr/bin/env python3

import argparse, os, json, pickle
import pandas as pd

def get_unified_funcname(funcname):
    if len(funcname) > 0:
        if '.' == funcname[0]:
            funcname = funcname[1:]
    return funcname

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("csv", help="input csv (e.g., Dataset-1)")
    ap.add_argument("pkl", help="input pkl generated by HermesSim")
    ap.add_argument("out", help="output json file")
    args = ap.parse_args()

    df = pd.read_csv(args.csv)
    pkl_data = pickle.load(open(args.pkl, 'rb'))
    assert len(df) == pkl_data.shape[0], "The number of functions not match"

    vecs = pkl_data.tolist()
    # e.g., IDBs/Dataset-1/z3/mips32-clang-3.5-O2_z3.i64
    bin_names = df["idb_path"].str.split('/').str[-1].str[:-4]
    func_names = df["func_name"].apply(get_unified_funcname)
    keys = bin_names + "|||" + func_names
    jd = {k: v for k, v in zip(keys, vecs)}

    os.makedirs(os.path.dirname(args.out), exist_ok=True)
    with open(args.out, "w") as fp:
        json.dump(jd, fp)
    print("Saved:", args.out)

if __name__ == "__main__":
    main()
```

**Comment by wang-yongpan** [comment]
We changed the save format by rewriting the original script of HermesSim.

Your script seems to be ok. 

You can try it first and contact us if you have any problems.

**Comment by TakahiroHaruyama** [comment]
Thanks!

**Comment by yangtt57** [comment]
@TakahiroHaruyama Hello, I've recently been trying to use HermesSim to generate initial embeddings as well, but I’ve encountered some difficulties. I noticed that you’ve worked on related topics, so I’d really appreciate your insights if you have a moment:
1. How can we obtain embeddings for auxiliary functions? From what I understand, HermesSim seems to train and test only on the selected (target) functions.

2.  What kind of data did you use to train the SEM module? Was it based on the validation set, or something else?

Thanks in advance for your help, and I really appreciate any guidance you can share!

**Comment by TakahiroHaruyama** [comment]
> How can we obtain embeddings for auxiliary functions? From what I understand, HermesSim seems to train and test only on the selected (target) functions.

In Dataset-1 case, the functions were filtered by a script provided by [binary_function_similarity](https://github.com/Cisco-Talos/binary_function_similarity/blob/main/DBs/Dataset-1/Dataset-1%20creation.ipynb). See the Step 1. 

> What kind of data did you use to train the SEM module? Was it based on the validation set, or something else?

For now, I've not retrained the model. I loaded the trained parameters from the Eval data.
